src/test/groovy/graphql/schema/idl/SchemaPrinterTest.groovy

```diff
@@ -2247,6 +2264,234 @@ type Query {
+                                newArgument().name("with").type(gunType)
+                                        .definition(newInputValueDefinition().comments(makeComments(" arg 'with'")).build()).build()))
+                                .build())
+                        .build())
+                .query(queryType)
+                .build()
+        when:
+        def result = new SchemaPrinter(defaultOptions().includeSchemaDefinition(true).includeAstDefinitionComments(true)).print(schema)
+        println(result)
+
+        then:
+        result == 
+'''#schema comment 1
+#       schema comment 2 with leading spaces
+schema {
+  query: Query
+  mutation: Mutation
+}
+
+"Marks the field, argument, input field or enum value as deprecated"
+directive @deprecated(
+    "The reason for the deprecation"
+    reason: String = "No longer supported"
+  ) on FIELD_DEFINITION | ARGUMENT_DEFINITION | ENUM_VALUE | INPUT_FIELD_DEFINITION
+
+" custom directive 'example' description 1"
+# custom directive 'example' comment 1
+directive @example on ENUM_VALUE
+
+"Directs the executor to include this field or fragment only when the `if` argument is true"
+directive @include(
+    "Included when true."
+    if: Boolean!
+  ) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT
+
+"Directs the executor to skip this field or fragment when the `if` argument is true."
+directive @skip(
+    "Skipped when true."
+    if: Boolean!
+  ) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT
+
+"Exposes a URL that specifies the behaviour of this scalar."
+directive @specifiedBy(
+    "The URL that specifies the behaviour of this scalar."
+    url: String!
+  ) on SCALAR
+
+# interface Character comment 1
+# interface Character comment 2
+interface Character implements Node {
+  appearsIn: [Episode]
+  friends: [Character]
+  id: ID!
+  name: String
+}
+
+interface Node {
+  id: ID!
+}
+
+# union type Humanoid comment 1
+union Humanoid = Droid | Human
+
+type Droid implements Character & Node {
+  appearsIn: [Episode]!
+  friends: [Character]
+  id: ID!
+  madeOn: Planet
+  name: String!
+  primaryFunction: String
+}
+
+type Human implements Character & Node {
+  appearsIn: [Episode]!
+  friends: [Character]
+  homePlanet: String
+  id: ID!
+  name: String!
+}
+
+type Mutation {
+  shoot(
+    # arg 'id\'
+    id: String!,
+    # arg 'with\'
+    with: Gun
+  ): Query
+}
+
+type Planet {
+  hitBy: Asteroid
+  name: String
+}
+
+# type query comment 1
+# type query comment 2
+type Query {
+  # query field 'hero' comment
+  hero(episode: Episode): Character
+  # query field 'humanoid' comment
+  humanoid(id: ID!): Humanoid
+}
+
+# enum Episode comment 1
+# enum Episode comment 2
+enum Episode {
+  # enum value EMPIRE comment 1
+  EMPIRE
+  JEDI
+  NEWHOPE @example
+}
+
+"desc"
+# scalar Asteroid comment 1
+scalar Asteroid
+
+# input type Gun comment 1
+input Gun {
+  # gun 'caliber' input value comment
+  caliber: Int
+  # gun 'name' input value comment
+  name: String
+}
 '''
     }
 }
```

Comment:
`graphql.schema.idl.SchemaGenerator.Options#useCommentsAsDescriptions` exists so it would be nice to test that we can generate from SDL with comments AND round drip it back out via a schema printer

This may well discover bugs if we have them in the SchemaGenerator

Reply:
Good call! I've added a test for this. Looks like it's not printing comments for directive and scalar. Will take a look at that later

Reply:
From what I can tell, the schema parser is not capturing directive and scalar comments in [getComments](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/parser/GraphqlAntlrToLanguage.java#L877) - the tokens seem to be ignored. I am not very familiar Antlr and don't see anything different for these two schema elements vs. all the other that seem to work fine. Any ideas?

Reply:
I dont off hand - what I would do is leave the SDL test in place BUT change the assertion string to leave our directives and scalars.  We can circle back to fix SDL scalar comment capture say later but this code can progress in the meantime